File: apps/nestjs-service/src/agent-for-test-case/testcase/shared/tools/export.ts
/**
 * Export tools for TestCase Agent (TypeScript port of shared/tools/export.py).
 *
 * Provides export functionality for test cases.
 * Note: These are utility functions called by the controller, not Agent tools.
 *
 * Excel export requires the 'xlsx' package:
 *   npm install xlsx
 * If not installed, JSON export is used as fallback.
 */

import * as fs from 'fs';
import * as path from 'path';

export type ExportFormat = 'excel' | 'xmind' | 'json';

/**
 * 导出测试用例（仅由 Controller 调用，非 Agent 工具）
 *
 * @param testCases  测试用例列表
 * @param format     导出格式（excel/xmind/json）
 * @param filename   文件名（不含扩展名）
 * @returns 导出文件路径
 */
export async function exportTestcases(
  testCases: Record<string, unknown>[],
  format: ExportFormat,
  filename: string,
): Promise<string> {
  if (!testCases || testCases.length === 0) {
    throw new Error('No test cases to export');
  }

  // 确保 exports 目录存在
  const exportsDir = path.resolve('exports');
  if (!fs.existsSync(exportsDir)) {
    fs.mkdirSync(exportsDir, { recursive: true });
  }

  switch (format) {
    case 'excel':
      return exportExcel(testCases, filename, exportsDir);
    case 'xmind':
      return exportXmind(testCases, filename, exportsDir);
    case 'json':
      return exportJson(testCases, filename, exportsDir);
    default:
      throw new Error(`Unsupported format: ${format}`);
  }
}

/**
 * 导出为 Excel
 *
 * Requires 'xlsx' package. Falls back to JSON if xlsx is not installed.
 */
async function exportExcel(
  testCases: Record<string, unknown>[],
  filename: string,
  exportsDir: string,
): Promise<string> {
  // 尝试动态 require xlsx（需要安装：npm install xlsx）
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let xlsx: any = null;
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    xlsx = require('xlsx');
  } catch {
    console.warn(
      '[Export] xlsx package not found, falling back to JSON. Install with: npm install xlsx',
    );
    return exportJson(testCases, `${filename}_excel_fallback`, exportsDir);
  }

  console.log(`[Export] Exporting ${testCases.length} test cases to Excel`);

  const headers = [
    '用例ID',
    '功能点ID',
    '标题',
    '类型',
    '优先级',
    '测试目的',
    '前置条件',
    '测试步骤',
    '预期结果',
    '标签',
  ];

  const rows: unknown[][] = [headers];

  for (const tc of testCases) {
    const steps = (tc.steps as Record<string, unknown>[] | undefined) ?? [];
    const stepsText = steps
      .map(
        (s, i) =>
          `${(s.step_no as number) ?? i + 1}. ${s.action ?? ''} (数据: ${s.test_data ?? ''})`,
      )
      .join('\n');

    const expectedResult = steps
      .map((s) => (s.expected_result as string) ?? '')
      .join('\n');

    const preconditions = ((tc.preconditions as string[] | undefined) ?? []).join('\n');
    const tags = ((tc.tags as string[] | undefined) ?? []).join(', ');

    rows.push([
      tc.id ?? '',
      tc.feature_point_id ?? '',
      tc.title ?? '',
      tc.type ?? '',
      tc.priority ?? '',
      tc.purpose ?? '',
      preconditions,
      stepsText,
      expectedResult,
      tags,
    ]);
  }

  const wb = xlsx.utils.book_new();
  const ws = xlsx.utils.aoa_to_sheet(rows);

  // 调整列宽
  ws['!cols'] = headers.map(() => ({ wch: 20 }));

  xlsx.utils.book_append_sheet(wb, ws, '测试用例');

  const outputPath = path.join(exportsDir, `${filename}.xlsx`);
  xlsx.writeFile(wb, outputPath);

  console.log(`[Export] Excel saved to: ${outputPath}`);
  return outputPath;
}

/**
 * 导出为 JSON
 */
async function exportJson(
  testCases: Record<string, unknown>[],
  filename: string,
  exportsDir: string,
): Promise<string> {
  console.log(`[Export] Exporting ${testCases.length} test cases to JSON`);

  const outputPath = path.join(exportsDir, `${filename}.json`);
  fs.writeFileSync(outputPath, JSON.stringify(testCases, null, 2), 'utf-8');

  console.log(`[Export] JSON saved to: ${outputPath}`);
  return outputPath;
}

/**
 * 导出为 XMind
 *
 * XMind 导出需要 xmind 库，目前使用 JSON 格式作为备选。
 */
async function exportXmind(
  testCases: Record<string, unknown>[],
  filename: string,
  exportsDir: string,
): Promise<string> {
  console.warn('[Export] XMind export not implemented, using JSON as fallback');
  return exportJson(testCases, `${filename}_xmind_fallback`, exportsDir);
}

File: src/agent-origin/testcase/shared/tools/export.py
"""Export tools for TestCase Agent.

This module provides export functionality for test cases.
Note: These are utility functions called by Controller, not Agent tools.
"""

import json
from typing import Literal
from pathlib import Path

from openpyxl import Workbook
from loguru import logger


async def export_testcases(
    test_cases: list[dict],
    format: Literal["excel", "xmind", "json"],
    filename: str,
) -> str:
    """
    导出测试用例（仅由 Controller 调用，非 Agent 工具）

    Args:
        test_cases: 测试用例列表
        format: 导出格式（excel/xmind/json）
        filename: 文件名（不含扩展名）

    Returns:
        导出文件路径

    Raises:
        ValueError: 不支持的格式
    """
    if not test_cases:
        raise ValueError("No test cases to export")

    # 确保 exports 目录存在
    exports_dir = Path("exports")
    exports_dir.mkdir(exist_ok=True)

    if format == "excel":
        return await _export_excel(test_cases, filename, exports_dir)
    elif format == "xmind":
        return await _export_xmind(test_cases, filename, exports_dir)
    elif format == "json":
        return await _export_json(test_cases, filename, exports_dir)
    else:
        raise ValueError(f"Unsupported format: {format}")


async def _export_excel(
    test_cases: list[dict], filename: str, exports_dir: Path
) -> str:
    """
    导出为 Excel

    Args:
        test_cases: 测试用例列表
        filename: 文件名
        exports_dir: 导出目录

    Returns:
        文件路径
    """
    logger.info(f"[Export] Exporting {len(test_cases)} test cases to Excel")

    wb = Workbook()
    ws = wb.active
    ws.title = "测试用例"

    # 表头
    headers = [
        "用例ID",
        "功能点ID",
        "标题",
        "类型",
        "优先级",
        "测试目的",
        "前置条件",
        "测试步骤",
        "预期结果",
        "标签",
    ]
    ws.append(headers)

    # 数据行
    for tc in test_cases:
        steps_text = "\n".join(
            [
                f"{s.get('step_no', i+1)}. {s.get('action', '')} "
                f"(数据: {s.get('test_data', '')})"
                for i, s in enumerate(tc.get("steps", []))
            ]
        )

        expected = "\n".join(
            [s.get("expected_result", "") for s in tc.get("steps", [])]
        )

        preconditions = "\n".join(tc.get("preconditions", []))
        tags = ", ".join(tc.get("tags", []))

        ws.append(
            [
                tc.get("id", ""),
                tc.get("feature_point_id", ""),
                tc.get("title", ""),
                tc.get("type", ""),
                tc.get("priority", ""),
                tc.get("purpose", ""),
                preconditions,
                steps_text,
                expected,
                tags,
            ]
        )

    # 调整列宽
    for column in ws.columns:
        max_length = 0
        column_letter = column[0].column_letter
        for cell in column:
            try:
                if len(str(cell.value)) > max_length:
                    max_length = len(str(cell.value))
            except:
                pass
        adjusted_width = min(max_length + 2, 50)
        ws.column_dimensions[column_letter].width = adjusted_width

    output_path = exports_dir / f"{filename}.xlsx"
    wb.save(output_path)

    logger.info(f"[Export] Excel saved to: {output_path}")
    return str(output_path)


async def _export_json(
    test_cases: list[dict], filename: str, exports_dir: Path
) -> str:
    """
    导出为 JSON

    Args:
        test_cases: 测试用例列表
        filename: 文件名
        exports_dir: 导出目录

    Returns:
        文件路径
    """
    logger.info(f"[Export] Exporting {len(test_cases)} test cases to JSON")

    output_path = exports_dir / f"{filename}.json"
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(test_cases, f, ensure_ascii=False, indent=2)

    logger.info(f"[Export] JSON saved to: {output_path}")
    return str(output_path)


async def _export_xmind(
    test_cases: list[dict], filename: str, exports_dir: Path
) -> str:
    """
    导出为 XMind

    Args:
        test_cases: 测试用例列表
        filename: 文件名
        exports_dir: 导出目录

    Returns:
        文件路径

    Note:
        XMind 导出需要 xmind 库，目前使用 JSON 格式作为备选
    """
    logger.warning(
        "[Export] XMind export not implemented, using JSON as fallback"
    )

    # 暂时使用 JSON 作为备选
    return await _export_json(test_cases, f"{filename}_xmind_fallback", exports_dir)
